File: experiments-local/experiment-16/runs/agent=omp_language=typescript_model=openrouter/moonshotai/kimi-k2.7-code_tooling=none/rep1/src/store.ts
/**
 * In-memory data store and query engine.
 *
 * All CSV rows are loaded once into memory. This keeps lookups fast (<2s) and
 * lets the MCP tools focus on formatting while the store answers the actual
 * soccer questions.
 */

import type { DataStore, HeadToHead, Match, Player, TeamRecord } from "./types.js";
import { normalizeCompetition, parseDate, teamDisplay, teamKey } from "./normalize.js";

export interface MatchFilters {
  team?: string;
  opponent?: string;
  from?: string | Date | null;
  to?: string | Date | null;
  competition?: string;
  season?: number;
  round?: string;
  venue?: "home" | "away" | "all";
  limit?: number;
}

export interface TeamStatFilters {
  season?: number;
  competition?: string;
  venue?: "home" | "away" | "all";
}

export interface PlayerFilters {
  name?: string;
  nationality?: string;
  club?: string;
  position?: string;
  minOverall?: number;
  limit?: number;
}

function toDate(value: string | Date | null | undefined): Date | null {
  if (value instanceof Date) return value;
  if (typeof value === "string" && value !== "") return parseDate(value);
  return null;
}

function dateCompare(a: Date | null, b: Date | null): number {
  if (a && b) return b.getTime() - a.getTime();
  if (a) return -1;
  if (b) return 1;
  return 0;
}

function matchesCompetition(match: Match, competition: string): boolean {
  return match.competition.toLowerCase().includes(competition.toLowerCase().trim());
}

export class SoccerStore {
  readonly matches: Match[];
  readonly players: Player[];
  private readonly teamDisplays = new Map<string, string>();

  constructor(data: DataStore) {
    this.matches = data.matches;
    this.players = data.players;
    this.indexTeamDisplays();
  }

  private indexTeamDisplays(): void {
    for (const match of this.matches) {
      if (!this.teamDisplays.has(match.homeKey)) {
        this.teamDisplays.set(match.homeKey, match.homeTeam);
      }
      if (!this.teamDisplays.has(match.awayKey)) {
        this.teamDisplays.set(match.awayKey, match.awayTeam);
      }
    }
    for (const player of this.players) {
      if (player.clubKey && !this.teamDisplays.has(player.clubKey)) {
        this.teamDisplays.set(player.clubKey, teamDisplay(player.club, player.clubKey));
      }
    }
  }

  displayFor(key: string): string {
    return this.teamDisplays.get(key) ?? teamDisplay("", key);
  }

  searchMatches(filters: MatchFilters): Match[] {
    const teamKeyFilter = filters.team ? teamKey(filters.team) : undefined;
    const opponentKey = filters.opponent ? teamKey(filters.opponent) : undefined;
    const from = toDate(filters.from);
    const to = toDate(filters.to);
    const competition = filters.competition ? normalizeCompetition(filters.competition) : undefined;
    const round = filters.round?.toLowerCase();

    const results = this.matches.filter((m) => {
      if (teamKeyFilter) {
        const isHome = m.homeKey === teamKeyFilter;
        const isAway = m.awayKey === teamKeyFilter;
        if (!isHome && !isAway) return false;
        if (filters.venue === "home" && !isHome) return false;
        if (filters.venue === "away" && !isAway) return false;
      }
      if (opponentKey) {
        if (m.homeKey !== opponentKey && m.awayKey !== opponentKey) return false;
        if (teamKeyFilter) {
          const hasBoth =
            (m.homeKey === teamKeyFilter && m.awayKey === opponentKey) ||
            (m.homeKey === opponentKey && m.awayKey === teamKeyFilter);
          if (!hasBoth) return false;
        }
      }
      if (competition && !matchesCompetition(m, competition)) return false;
      if (filters.season !== undefined && m.season !== filters.season) return false;
      if (from && m.date && m.date < from) return false;
      if (to && m.date && m.date > to) return false;
      if (round && (!m.round || !m.round.toLowerCase().includes(round))) return false;
      return true;
    });

    results.sort((a, b) => dateCompare(a.date, b.date));
    return filters.limit ? results.slice(0, filters.limit) : results;
  }

  teamStatistics(team: string, filters: TeamStatFilters = {}): TeamRecord {
    const key = teamKey(team);
    const venue = filters.venue ?? "all";
    const competition = filters.competition ? normalizeCompetition(filters.competition) : undefined;

    let matches = 0;
    let wins = 0;
    let draws = 0;
    let losses = 0;
    let goalsFor = 0;
    let goalsAgainst = 0;

    for (const m of this.matches) {
      if (filters.season !== undefined && m.season !== filters.season) continue;
      if (competition && !matchesCompetition(m, competition)) continue;

      const isHome = m.homeKey === key;
      const isAway = m.awayKey === key;
      if (!isHome && !isAway) continue;
      if (venue === "home" && !isHome) continue;
      if (venue === "away" && !isAway) continue;
      if (m.homeGoals === null || m.awayGoals === null) continue;

      matches++;
      const forGoals = isHome ? m.homeGoals : m.awayGoals;
      const againstGoals = isHome ? m.awayGoals : m.homeGoals;
      goalsFor += forGoals;
      goalsAgainst += againstGoals;

      if (forGoals > againstGoals) wins++;
      else if (forGoals === againstGoals) draws++;
      else losses++;
    }

    return {
      team: this.displayFor(key),
      key,
      matches,
      wins,
      draws,
      losses,
      goalsFor,
      goalsAgainst,
      points: wins * 3 + draws,
    };
  }

  headToHead(teamA: string, teamB: string, filters: Omit<TeamStatFilters, "venue"> = {}): HeadToHead {
    const keyA = teamKey(teamA);
    const keyB = teamKey(teamB);
    const competition = filters.competition ? normalizeCompetition(filters.competition) : undefined;

    const matches = this.matches.filter((m) => {
      const hasA = m.homeKey === keyA || m.awayKey === keyA;
      const hasB = m.homeKey === keyB || m.awayKey === keyB;
      if (!hasA || !hasB) return false;
      if (filters.season !== undefined && m.season !== filters.season) return false;
      if (competition && !matchesCompetition(m, competition)) return false;
      if (m.homeGoals === null || m.awayGoals === null) return false;
      return true;
    });

    matches.sort((a, b) => dateCompare(a.date, b.date));

    let winsA = 0;
    let winsB = 0;
    let draws = 0;
    for (const m of matches) {
      const aIsHome = m.homeKey === keyA;
      const aGoals = aIsHome ? m.homeGoals! : m.awayGoals!;
      const bGoals = aIsHome ? m.awayGoals! : m.homeGoals!;
      if (aGoals > bGoals) winsA++;
      else if (aGoals < bGoals) winsB++;
      else draws++;
    }

    return {
      teamA: this.displayFor(keyA),
      teamB: this.displayFor(keyB),
      matches,
      winsA,
      winsB,
      draws,
    };
  }

  competitionStandings(competition: string, season: number): TeamRecord[] {
    const norm = normalizeCompetition(competition);
    const table = new Map<string, TeamRecord>();

    for (const m of this.matches) {
      if (m.season !== season) continue;
      if (!matchesCompetition(m, norm)) continue;
      if (m.homeGoals === null || m.awayGoals === null) continue;

      this.updateStanding(table, m.homeKey, m.homeGoals, m.awayGoals);
      this.updateStanding(table, m.awayKey, m.awayGoals, m.homeGoals);
    }

    return Array.from(table.values()).sort((a, b) => {
      if (b.points !== a.points) return b.points - a.points;
      const diffA = a.goalsFor - a.goalsAgainst;
      const diffB = b.goalsFor - b.goalsAgainst;
      if (diffB !== diffA) return diffB - diffA;
      return b.goalsFor - a.goalsFor;
    });
  }

  private updateStanding(
    table: Map<string, TeamRecord>,
    key: string,
    forGoals: number,
    againstGoals: number
  ): void {
    const existing = table.get(key);
    const record = existing ?? {
      team: this.displayFor(key),
      key,
      matches: 0,
      wins: 0,
      draws: 0,
      losses: 0,
      goalsFor: 0,
      goalsAgainst: 0,
      points: 0,
    };
    record.matches++;
    record.goalsFor += forGoals;
    record.goalsAgainst += againstGoals;
    if (forGoals > againstGoals) {
      record.wins++;
      record.points += 3;
    } else if (forGoals === againstGoals) {
      record.draws++;
      record.points += 1;
    } else {
      record.losses++;
    }
    if (!existing) table.set(key, record);
  }

  biggestWins(filters: Omit<MatchFilters, "team" | "opponent" | "venue"> = {}): Match[] {
    const competition = filters.competition ? normalizeCompetition(filters.competition) : undefined;
    const from = toDate(filters.from);
    const to = toDate(filters.to);

    const results = this.matches.filter((m) => {
      if (filters.season !== undefined && m.season !== filters.season) return false;
      if (competition && !matchesCompetition(m, competition)) return false;
      if (from && m.date && m.date < from) return false;
      if (to && m.date && m.date > to) return false;
      return m.homeGoals !== null && m.awayGoals !== null;
    });

    results.sort((a, b) => {
      const diffA = Math.abs((a.homeGoals ?? 0) - (a.awayGoals ?? 0));
      const diffB = Math.abs((b.homeGoals ?? 0) - (b.awayGoals ?? 0));
      return diffB - diffA;
    });

    return filters.limit ? results.slice(0, filters.limit) : results;
  }

  searchPlayers(filters: PlayerFilters): Player[] {
    const clubKey = filters.club ? teamKey(filters.club) : undefined;
    const nameLower = filters.name?.toLowerCase().trim();
    const nationalityLower = filters.nationality?.toLowerCase().trim();
    const positionUpper = filters.position?.toUpperCase().trim();

    const results = this.players.filter((p) => {
      if (nameLower && !p.name.toLowerCase().includes(nameLower)) return false;
      if (nationalityLower && !p.nationality?.toLowerCase().includes(nationalityLower)) return false;
      if (clubKey && p.clubKey !== clubKey) return false;
      if (positionUpper && p.position?.toUpperCase() !== positionUpper) return false;
      if (filters.minOverall !== undefined && (p.overall ?? 0) < filters.minOverall) return false;
      return true;
    });

    results.sort((a, b) => (b.overall ?? 0) - (a.overall ?? 0));
    return filters.limit ? results.slice(0, filters.limit) : results;
  }

  averageGoalsPerMatch(filters: Omit<MatchFilters, "team" | "opponent" | "venue"> = {}): number {
    const competition = filters.competition ? normalizeCompetition(filters.competition) : undefined;
    let total = 0;
    let count = 0;
    for (const m of this.matches) {
      if (filters.season !== undefined && m.season !== filters.season) continue;
      if (competition && !matchesCompetition(m, competition)) continue;
      if (m.homeGoals === null || m.awayGoals === null) continue;
      total += m.homeGoals + m.awayGoals;
      count++;
    }
    return count === 0 ? 0 : total / count;
  }
}

File: experiments-local/experiment-mu-glm53-brazil/runs/agent=opencode_language=python_model=openrouter/z-ai/glm-5.3-flash_tooling=none/rep3/brazilian_soccer/store.py
"""In-memory knowledge store: indexes over matches/players + queries.

``SoccerStore`` is the single query surface shared by the MCP tools and
the tests.  It loads the datasets once (about 0.6 s for ~17k matches and
18k players) and keeps O(1) lookup indexes:

- ``matches_by_team``      canonical team key -> matches it played
- ``matches_by_comp_season`` (competition, season) -> matches
- ``matches_by_stage``     (competition, stage) -> matches (finals etc.)
- ``players_by_club``      canonical club key -> players
- ``players_sorted``       all players sorted by FIFA Overall

Fuzzy resolution helpers make natural-language-style arguments work:
``resolve_team("Flamengo")`` and ``resolve_team("Flamengo-RJ")`` return
the same canonical key; ``resolve_competition("brasileirao")`` returns
``Brasileirão Serie A``.  Traditional rivalries (derbies) are defined in
``DERBIES`` and drive ``derbies()`` and derby detection.
"""

from __future__ import annotations

import re
from collections import defaultdict
from datetime import date

from .analytics import (
    aggregate_stats,
    best_venues,
    biggest_wins,
    h2h_record,
    season_comparison,
    standings_table,
    team_record,
)
from .loader import load_all
from .models import (
    COPA_DO_BRASIL,
    LEAGUE_COMPETITIONS,
    LIBERTADORES,
    BRASILEIRAO_A,
    Match,
    Player,
)
from .normalize import parse_date, strip_accents

DEFAULT_LIMIT = 25
MAX_LIMIT = 100

# Traditional rivalries: (derby name, aliases, raw team names).
DERBIES: list[tuple[str, list[str], tuple[str, str]]] = [
    ("Fla-Flu", ["fla flu", "fla-flu"], ("Flamengo-RJ", "Fluminense-RJ")),
    ("Clássico dos Milhões", ["classico dos milhoes"],
     ("Flamengo-RJ", "Vasco da Gama-RJ")),
    ("Clássico da Rivalidade", ["classico da rivalidade"],
     ("Botafogo-RJ", "Flamengo-RJ")),
    ("Clássico Vovô", ["classico vovo"], ("Botafogo-RJ", "Fluminense-RJ")),
    ("Dérbi Paulista", ["derbi paulista", "derby paulista"],
     ("Corinthians-SP", "Palmeiras-SP")),
    ("Majestoso", ["majestoso"], ("Corinthians-SP", "São Paulo-SP")),
    ("Choque-Rei", ["choque rei", "choque-rei"], ("Palmeiras-SP", "São Paulo-SP")),
    ("San-São", ["san sao"], ("Santos-SP", "São Paulo-SP")),
    ("Grenal", ["grenal"], ("Grêmio-RS", "Internacional-RS")),
    ("Atletiba", ["atletiba"], ("Athletico-PR", "Coritiba-PR")),
    ("Ba-Vi", ["ba vi", "ba-vi"], ("Bahia-BA", "Vitória-BA")),
    ("Clássico Mineiro", ["classico mineiro"], ("Atlético-MG", "Cruzeiro-MG")),
    ("Clássico-Rei", ["classico rei"], ("Ceará-CE", "Fortaleza-CE")),
    ("Clássico das Multidões", ["classico das multidoes"],
     ("Sport-PE", "Santa Cruz-PE")),
    ("Clássico Emoção", ["classico emocao"], ("Sport-PE", "Náutico-PE")),
]

COMPETITION_ALIASES = {
    "brasileirao": BRASILEIRAO_A,
    "brasileirao serie a": BRASILEIRAO_A,
    "serie a": BRASILEIRAO_A,
    "brasileirao serie b": "Brasileirão Serie B",
    "serie b": "Brasileirão Serie B",
    "brasileirao serie c": "Brasileirão Serie C",
    "serie c": "Brasileirão Serie C",
    "copa do brasil": COPA_DO_BRASIL,
    "brazilian cup": COPA_DO_BRASIL,
    "copa brasil": COPA_DO_BRASIL,
    "libertadores": LIBERTADORES,
    "copa libertadores": LIBERTADORES,
}


class NotFound(LookupError):
    """Raised when a team/competition/player cannot be resolved."""


class SoccerStore:
    """Query engine over the six Brazilian soccer datasets."""

    def __init__(self) -> None:
        self.registry, self.matches, self.players = load_all()
        self._build_indexes()

    # ------------------------------------------------------------------
    # indexes
    # ------------------------------------------------------------------

    def _build_indexes(self) -> None:
        self.matches_by_team: dict[str, list[Match]] = defaultdict(list)
        self.matches_by_comp_season: dict[tuple[str, int], list[Match]] = defaultdict(list)
        self.matches_by_stage: dict[tuple[str, str], list[Match]] = defaultdict(list)
        for m in self.matches:
            self.matches_by_team[m.home_key].append(m)
            self.matches_by_team[m.away_key].append(m)
            self.matches_by_comp_season[(m.competition, m.season)].append(m)
            if m.stage:
                self.matches_by_stage[(m.competition, m.stage)].append(m)
        self.players_by_club: dict[str, list[Player]] = defaultdict(list)
        for p in self.players:
            if p.club_key:
                self.players_by_club[p.club_key].append(p)
        self.players_sorted = sorted(self.players, key=lambda p: -p.overall)

        # Derby pairs in canonical key space.
        self.derby_pairs: list[tuple[str, tuple[str, str]]] = []
        for name, _aliases, (a, b) in DERBIES:
            self.derby_pairs.append(
                (name, (self.registry.key_for(a), self.registry.key_for(b))))

        # Team popularity: matches played (used to pick fuzzy best match).
        self._team_popularity: dict[str, int] = {
            k: len(v) for k, v in self.matches_by_team.items()}

    # ------------------------------------------------------------------
    # resolution helpers
    # ------------------------------------------------------------------

    def resolve_team(self, name: str) -> str:
        """Canonical key for a team name; fuzzy-falls back to displays."""
        if not name or not name.strip():
            raise NotFound("Empty team name")
        raw = name.strip()
        key = self.registry.key_for(raw)
        if key in self._team_popularity or key in self.players_by_club:
            return key
        # substring match over display names (accent-insensitive)
        q = strip_accents(raw).lower()
        cands = {k: d for k, d in self.registry.display_names().items()
                 if q in strip_accents(d).lower()}
        if not cands:
            raise NotFound(
                f"Team {name!r} not found; dataset covers "
                f"{len(self.registry.display_names())} clubs")
        # prefer the candidate with the most matches (famous club wins)
        return max(cands, key=lambda k: self._team_popularity.get(k, 0))

    def resolve_competition(self, name: str | None) -> str | None:
        """Canonical competition label (None matches everything)."""
        if name is None or not str(name).strip():
            return None
        q = strip_accents(str(name)).strip().lower()
        if q in COMPETITION_ALIASES:
            return COMPETITION_ALIASES[q]
        all_names = set(COMPETITION_ALIASES.values())
        for label in all_names:
            if q in strip_accents(label).lower():
                return label
        raise NotFound(
            f"Competition {name!r} not found; known: {sorted(all_names)}")

    def team_display(self, key: str) -> str:
        return self.registry.display(key)

    def derby_of(self, home_key: str, away_key: str) -> str | None:
        """Derby name if the fixture is a traditional rivalry."""
        for name, (a, b) in self.derby_pairs:
            if {home_key, away_key} == {a, b}:
                return name
        return None

    # ------------------------------------------------------------------
    # generic filters
    # ------------------------------------------------------------------

    @staticmethod
    def _filters_apply(m: Match, competition: str | None, season: int | None,
                       date_from, date_to) -> bool:
        if competition and m.competition != competition:
            return False
        if season is not None and m.season != season:
            return False
        if date_from and (m.date is None or m.date < date_from):
            return False
        if date_to and (m.date is None or m.date > date_to):
            return False
        return True

    @staticmethod
    def match_dict(m: Match, store: "SoccerStore") -> dict:
        """JSON-ready representation of a match."""
        home = store.registry.display(m.home_key)
        away = store.registry.display(m.away_key)
        return {
            "date": m.date.isoformat() if m.date else None,
            "time": m.time,
            "competition": m.competition,
            "season": m.season,
            "round": m.round,
            "stage": m.stage,
            "home": home,
            "away": away,
            "home_goal": m.home_goal,
            "away_goal": m.away_goal,
            "score": (f"{home} {m.home_goal}-{m.away_goal} {away}"
                      if m.has_result() else None),
            "winner": (store.registry.display(m.winner_key())
                       if m.winner_key() else ("Draw" if m.has_result() else None)),
            "arena": m.arena,
            "source": m.source,
        }

    # ------------------------------------------------------------------
    # match queries
    # ------------------------------------------------------------------

    def search_matches(self, team: str | None = None,
                       opponent: str | None = None,
                       competition: str | None = None,
                       season: int | None = None,
                       stage: str | None = None,
                       date_from: str | None = None,
                       date_to: str | None = None,
                       limit: int = DEFAULT_LIMIT,
                       order: str = "asc") -> dict:
        comp = self.resolve_competition(competition)
        season = _as_int(season)
        d_from = parse_date(date_from) if date_from else None
        d_to = parse_date(date_to) if date_to else None
        stage_q = strip_accents(stage).lower() if stage else None

        team_key = self.resolve_team(team) if team else None
        opp_key = self.resolve_team(opponent) if opponent else None

        if team_key and opp_key:
            pool = [m for m in self.matches_by_team[team_key]
                    if opp_key in (m.home_key, m.away_key)]
        elif team_key:
            pool = self.matches_by_team[team_key]
        elif opp_key:
            pool = self.matches_by_team[opp_key]
        else:
            pool = self.matches

        selected = [
            m for m in pool
            if self._filters_apply(m, comp, season, d_from, d_to)
            and (not stage_q or (m.stage or "").lower() == stage_q)
        ]
        selected.sort(key=lambda m: (m.date is None, m.date),
                      reverse=(order == "desc"))
        return {
            "total": len(selected),
            "matches": [self.match_dict(m, self) for m in selected[:_clamp(limit)]],
        }

    def head_to_head(self, team_a: str, team_b: str,
                     competition: str | None = None,
                     limit: int = DEFAULT_LIMIT) -> dict:
        a = self.resolve_team(team_a)
        b = self.resolve_team(team_b)
        if a == b:
            raise NotFound("Head-to-head needs two different teams")
        comp = self.resolve_competition(competition)
        pool = [m for m in self.matches_by_team[a] if b in (m.home_key, m.away_key)]
        selected = [m for m in pool if self._filters_apply(m, comp, None, None, None)]
        selected.sort(key=lambda m: (m.date is None, m.date))
        rec = h2h_record(selected, a, b)
        return {
            "team_a": self.team_display(a),
            "team_b": self.team_display(b),
            "competition": comp,
            "team_a_wins": rec["team_a_wins"],
            "team_b_wins": rec["team_b_wins"],
            "draws": rec["draws"],
            "team_a_goals": rec["team_a_goals"],
            "team_b_goals": rec["team_b_goals"],
            "derby": self.derby_of(a, b),
            "total_matches": len(selected),
            "total": len(selected),          # alias shared with search_matches
            "matches": [self.match_dict(m, self) for m in selected[-_clamp(limit):]],
        }

    # ------------------------------------------------------------------
    # team queries
    # ------------------------------------------------------------------

    def team_stats(self, team: str, season: int | None = None,
                   competition: str | None = None,
                   venue: str | None = None) -> dict:
        key = self.resolve_team(team)
        comp = self.resolve_competition(competition)
        season = _as_int(season)
        pool = [m for m in self.matches_by_team[key]
                if self._filters_apply(m, comp, season, None, None)]
        venue_n = (venue or "").strip().lower()
        if venue_n in ("home", "away"):
            pool = [m for m in pool
                    if (m.home_key == key) == (venue_n == "home")]
        rec = team_record(pool, key)
        return {
            "team": self.team_display(key),
            "competition": comp,
            "season": season,
            "venue": venue_n or "all",
            **rec,
        }

    def team_history(self, team: str) -> dict:
        key = self.resolve_team(team)
        per_comp: dict[str, set[int]] = defaultdict(set)
        for m in self.matches_by_team[key]:
            per_comp[m.competition].add(m.season)
        return {
            "team": self.team_display(key),
            "total_matches": len(self.matches_by_team[key]),
            "overall_record": team_record(self.matches_by_team[key], key),
            "competitions": {
                comp: sorted(seasons) for comp, seasons in sorted(per_comp.items())
            },
        }

    def list_teams(self, competition: str | None = None,
                   season: int | None = None) -> dict:
        comp = self.resolve_competition(competition)
        season = _as_int(season)
        teams = {}
        for m in self.matches:
            if not self._filters_apply(m, comp, season, None, None):
                continue
            for k in (m.home_key, m.away_key):
                teams[k] = teams.get(k, 0) + 1
        return {
            "total": len(teams),
            "teams": [
                {"team": self.team_display(k), "matches": n}
                for k, n in sorted(teams.items(), key=lambda kv: -kv[1])
            ],
        }

    # ------------------------------------------------------------------
    # competition queries
    # ------------------------------------------------------------------

    def competitions(self) -> dict:
        catalog = defaultdict(lambda: {"seasons": set(), "matches": 0})
        for m in self.matches:
            entry = catalog[m.competition]
            entry["seasons"].add(m.season)
            entry["matches"] += 1
        return {
            "competitions": [
                {
                    "competition": comp,
                    "seasons": f"{min(ss)}-{max(ss)}" if ss else "-",
                    "season_count": len(ss),
                    "matches": e["matches"],
                }
                for comp, e in sorted(catalog.items())
                for ss in [e["seasons"]]
            ]
        }

    def standings(self, competition: str, season: int,
                  top: int | None = None) -> dict:
        comp = self.resolve_competition(competition)
        season = _as_int(season)
        if season is None:
            raise NotFound("Standings need a season (year)")
        if comp not in LEAGUE_COMPETITIONS:
            raise NotFound(
                f"Standings are only computed for league competitions "
                f"({sorted(LEAGUE_COMPETITIONS)}), not {comp!r}")
        matches = self.matches_by_comp_season.get((comp, season), [])
        table = standings_table(matches)
        if top:
            table = table[:top]
        rows = []
        for r in table:
            rows.append({
                "position": r["position"],
                "team": self.team_display(r["team"]),
                "matches": r["matches"],
                "wins": r["wins"],
                "draws": r["draws"],
                "losses": r["losses"],
                "goals_for": r["goals_for"],
                "goals_against": r["goals_against"],
                "goal_difference": r["goal_difference"],
                "points": r["points"],
            })
        response = {
            "competition": comp,
            "season": season,
            "total_matches_used": len(matches),
            "table": rows,
            "champion": rows[0]["team"] if rows else None,
        }
        if comp == BRASILEIRAO_A and len(rows) >= 4:
            response["relegation_zone"] = [r["team"] for r in rows[-4:]]
        return response

    # ------------------------------------------------------------------
    # player queries
    # ------------------------------------------------------------------

    @staticmethod
    def _player_dict(p: Player, with_skills: bool = False) -> dict:
        d = {
            "id": p.id,
            "name": p.name,
            "nationality": p.nationality,
            "overall": p.overall,
            "potential": p.potential,
            "club": p.club_display,
            "position": p.position,
            "age": p.age,
            "preferred_foot": p.preferred_foot,
        }
        if with_skills:
            d.update({
                "height_cm": p.height_cm,
                "weight_kg": p.weight_kg,
                "jersey": p.jersey,
                "value": p.value,
                "wage": p.wage,
                "skills": p.skills,
            })
        return d

    def search_players(self, name: str | None = None,
                       nationality: str | None = None,
                       club: str | None = None,
                       position: str | None = None,
                       min_overall: int | None = None,
                       max_overall: int | None = None,
                       limit: int = DEFAULT_LIMIT) -> dict:
        club_key = None
        if club:
            try:
                club_key = self.resolve_team(club)
            except NotFound:
                club_key = None
        name_q = strip_accents(name or "").lower()
        nat_q = (nationality or "").strip().lower()
        pos_q = (position or "").strip().upper()
        min_o = _as_int(min_overall)
        max_o = _as_int(max_overall)

        hits = []
        for p in self.players_sorted:
            if club_key and p.club_key != club_key:
                continue
            if nat_q and nat_q not in p.nationality.lower():
                continue
            if pos_q and p.position != pos_q:
                continue
            if min_o is not None and p.overall < min_o:
                continue
            if max_o is not None and p.overall > max_o:
                continue
            if name_q and name_q not in strip_accents(p.name).lower():
                continue
            hits.append(p)
        return {
            "total": len(hits),
            "players": [self._player_dict(p) for p in hits[:_clamp(limit)]],
        }

    def get_player(self, name: str) -> dict:
        q = strip_accents(name).strip().lower()
        if not q:
            raise NotFound("Empty player name")
        # exact normalized match first
        for p in self.players_sorted:
            if strip_accents(p.name).lower() == q:
                return self._player_dict(p, with_skills=True)
        # then token-set match ("gabriel barbosA" ~ "Gabriel Barbosa")
        q_tokens = set(re.split(r"[^a-z0-9]+", q)) - {""}
        best, best_score = None, 0
        for p in self.players_sorted:
            tokens = set(re.split(r"[^a-z0-9]+", strip_accents(p.name).lower())) - {""}
            if tokens and q_tokens and q_tokens <= tokens:
                score = len(tokens)
                if score > best_score:
                    best, best_score = p, score
        if best is None:
            partial = [p for p in self.players_sorted
                       if q in strip_accents(p.name).lower()]
            if partial:
                best = max(partial, key=lambda p: p.overall)
        if best is None:
            raise NotFound(f"Player {name!r} not found in the FIFA dataset")
        return self._player_dict(best, with_skills=True)

    def players_at_club(self, club: str, limit: int = DEFAULT_LIMIT) -> dict:
        key = self.resolve_team(club)
        members = self.players_by_club.get(key, [])
        members.sort(key=lambda p: -p.overall)
        avg = round(sum(p.overall for p in members) / len(members), 1) if members else 0.0
        response = {
            "club": self.team_display(key),
            "total_players": len(members),
            "total": len(members),           # alias shared with search_players
            "average_overall": avg,
            "players": [self._player_dict(p) for p in members[:_clamp(limit)]],
        }
        if not members:
            response["note"] = (
                "The bundled FIFA dataset (FIFA 19 edition) does not include "
                "squad data for every Brazilian club; try clubs like "
                "Cruzeiro, Santos, Fluminense, Grêmio or Atlético Mineiro.")
        return response

    # ------------------------------------------------------------------
    # statistics
    # ------------------------------------------------------------------

    def statistics(self, competition: str | None = None,
                   season: int | None = None) -> dict:
        comp = self.resolve_competition(competition)
        season = _as_int(season)
        pool = [m for m in self.matches
                if self._filters_apply(m, comp, season, None, None)]
        stats = aggregate_stats(pool)
        return {
            "scope": {"competition": comp, "season": season},
            **stats,
            "biggest_wins": [
                {
                    "date": x["match"].date.isoformat() if x["match"].date else None,
                    "winner": self.team_display(x["winner"]),
                    "score": f"{x['winner_goals']}-{x['loser_goals']}",
                    "loser": self.team_display(x["loser"]),
                    "competition": x["match"].competition,
                }
                for x in biggest_wins(pool, top=10)
            ],
            "best_home_records": [
                {"team": self.team_display(x["team"]), **x["record"]}
                for x in best_venues(pool, "home", top=5)
            ],
            "best_away_records": [
                {"team": self.team_display(x["team"]), **x["record"]}
                for x in best_venues(pool, "away", top=5)
            ],
        }

    def compare_seasons(self, competition: str, season_a: int,
                        season_b: int) -> dict:
        comp = self.resolve_competition(competition)
        season_a, season_b = _as_int(season_a), _as_int(season_b)
        per = {s: self.matches_by_comp_season.get((comp, s), [])
               for s in (season_a, season_b)}
        return {
            "competition": comp,
            "seasons": season_comparison(per),
            "champions": {
                s: (self.team_display(standings_table(ms)[0]["team"])
                    if ms else None)
                for s, ms in per.items()
            },
        }

    # ------------------------------------------------------------------
    # derbies
    # ------------------------------------------------------------------

    def derbies(self, season: int | None = None,
                competition: str | None = None) -> dict:
        season = _as_int(season)
        comp = self.resolve_competition(competition)
        out = []
        for name, (a, b) in self.derby_pairs:
            pool = [m for m in self.matches_by_team[a]
                    if b in (m.home_key, m.away_key)
                    and self._filters_apply(m, comp, season, None, None)]
            if not pool:
                continue
            rec = h2h_record(pool, a, b)
            out.append({
                "derby": name,
                "team_a": self.team_display(a),
                "team_b": self.team_display(b),
                "team_a_wins": rec["team_a_wins"],
                "team_b_wins": rec["team_b_wins"],
                "draws": rec["draws"],
                "total_matches": len(pool),
                "recent": [self.match_dict(m, self)
                           for m in sorted(pool, key=lambda m: m.date or date.min)][-3:],
            })
        return {"season": season, "derbies": out}

    # ------------------------------------------------------------------
    # status / catalog
    # ------------------------------------------------------------------

    def status(self) -> dict:
        sources = defaultdict(int)
        for m in self.matches:
            sources[m.source] += 1
        return {
            "matches": len(self.matches),
            "players": len(self.players),
            "teams": len(self.registry.display_names()),
            "competitions": sorted({m.competition for m in self.matches}),
            "sources": dict(sorted(sources.items())),
        }


# ----------------------------------------------------------------------
# helpers
# ----------------------------------------------------------------------

def _as_int(value) -> int | None:
    if value is None or value == "":
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _clamp(limit) -> int:
    n = _as_int(limit) or DEFAULT_LIMIT
    return max(1, min(n, MAX_LIMIT))
